modules/order/src/OrderRefresh.php

```diff
@@ -98,41 +97,65 @@ public function needsRefresh(OrderInterface $order) {
-      $adjustments = $order->getAdjustments();
-      foreach ($adjustments as $key => $adjustment) {
-        if ($adjustment->getType() != 'custom') {
-          unset($adjustments[$key]);
-        }
-      }
-      $order->setAdjustments($adjustments);
+  public function needsRefresh(OrderInterface $order) {
+    // Only draft orders should be automatically refreshed.
+    if ($order->getState()->value != 'draft') {
+      return FALSE;
+    }
 
-      foreach ($order->getItems() as $order_item) {
-        $order_item->setAdjustments([]);
+    // Accommodate long-running processes by always using the current time.
+    $current_time = $this->time->getCurrentTime();
+    $order_time = $order->getChangedTime();
+    if (date('Y-m-d', $current_time) != date('Y-m-d', $order_time)) {
+      // Refresh on a date change regardless of the refresh frequency.
+      // Date changes can impact tax rate amounts, availability of promotions.
+      return TRUE;
+    }
+    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+    $order_type = $this->orderTypeStorage->load($order->bundle());
+    $refreshed_ago = $current_time - $order_time;
+    if ($refreshed_ago >= $order_type->getRefreshFrequency()) {
+      return TRUE;
+    }
 
-        $purchased_entity = $order_item->getPurchasedEntity();
-        // @todo resolve pricing for items without purchaseable entity.
-        if ($purchased_entity) {
-          $order_item->setTitle($purchased_entity->getOrderItemTitle());
-          $unit_price = $this->chainPriceResolver->resolve($purchased_entity, $order_item->getQuantity());
-          $order_item->setUnitPrice($unit_price);
-        }
+    return FALSE;
+  }
 
-        $order_item->save();
+  /**
+   * {@inheritdoc}
+   */
+  public function refresh(OrderInterface $order) {
+    // Do not remove adjustments added in the user interface.
+    $adjustments = $order->getAdjustments();
+    foreach ($adjustments as $key => $adjustment) {
+      if ($adjustment->getType() != 'custom') {
+        unset($adjustments[$key]);
       }
+    }
+    $order->setAdjustments($adjustments);
+
+    foreach ($order->getItems() as $order_item) {
+      $order_item->setAdjustments([]);
 
-      foreach ($this->processors as $processor) {
-        $processor->process($order);
+      $purchased_entity = $order_item->getPurchasedEntity();
+      if ($purchased_entity) {
+        $order_item->setTitle($purchased_entity->getOrderItemTitle());
+        $unit_price = $this->chainPriceResolver->resolve($purchased_entity, $order_item->getQuantity());
+        $order_item->setUnitPrice($unit_price);
       }
+    }
+
+    // Allow the processors to modify the order and its items.
+    foreach ($this->processors as $processor) {
+      $processor->process($order);
+    }
 
-      $order->save();
-      unset(self::$refreshingOrders[$order->id()]);
+    // @todo Evaluate which order items have changed.
+    foreach ($order->getItems() as $order_item) {
```

Comment:
If we did that then order processors would be useless for order items, which would be both confusing and limiting.